File: orcastork/adapters/memory/catalog.py
"""In-memory ``CapabilityCatalog`` — explicit permitted-set + credentials maps.

Enough to satisfy the port contract and the conformance suite; a deployment writes its own
adapter over its real configuration store. The mutators let tests model a namespace enabling/revoking a
capability between grants.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping, Sequence
from typing import Any

from ...ids import CapabilityId, NamespaceId, OperatorId


class InMemoryCapabilityCatalog:
    def __init__(
        self,
        permitted: Mapping[NamespaceId, Iterable[CapabilityId]] | None = None,
        credentials: Mapping[tuple[NamespaceId, CapabilityId], Mapping[str, Any]] | None = None,
        preferred: Mapping[NamespaceId, Sequence[CapabilityId]] | None = None,
        permitted_operators: Mapping[NamespaceId, Iterable[OperatorId]] | None = None,
    ) -> None:
        self._permitted: dict[NamespaceId, frozenset[CapabilityId]] = {
            namespace: frozenset(caps) for namespace, caps in (permitted or {}).items()
        }
        self._credentials: dict[tuple[NamespaceId, CapabilityId], Mapping[str, Any]] = {
            key: dict(value) for key, value in (credentials or {}).items()
        }
        self._preferred: dict[NamespaceId, tuple[CapabilityId, ...]] = {
            namespace: tuple(order) for namespace, order in (preferred or {}).items()
        }
        self._permitted_operators: dict[NamespaceId, frozenset[OperatorId]] = {
            namespace: frozenset(operators) for namespace, operators in (permitted_operators or {}).items()
        }

    async def permitted_capabilities(self, namespace_id: NamespaceId) -> frozenset[CapabilityId]:
        return self._permitted.get(namespace_id, frozenset())

    async def permitted_operators(self, namespace_id: NamespaceId) -> frozenset[OperatorId] | None:
        # A namespace with no configured restriction is unrestricted (None), per the port contract —
        # an empty frozenset is the explicit "run nothing" configuration, never the default.
        return self._permitted_operators.get(namespace_id)

    async def credentials(self, namespace_id: NamespaceId, capability_id: CapabilityId) -> Mapping[str, Any] | None:
        # Hand back a copy so a caller mutating the returned mapping cannot alter catalog state.
        stored = self._credentials.get((namespace_id, capability_id))
        return None if stored is None else dict(stored)

    async def preferred_order(self, namespace_id: NamespaceId) -> tuple[CapabilityId, ...]:
        return self._preferred.get(namespace_id, ())

    def set_permitted(self, namespace_id: NamespaceId, capabilities: Iterable[CapabilityId]) -> None:
        """Test helper: model a namespace config change (enable/revoke) visible to the next grant."""
        self._permitted[namespace_id] = frozenset(capabilities)

    def set_preferred_order(self, namespace_id: NamespaceId, order: Sequence[CapabilityId]) -> None:
        """Test helper: model a namespace changing its provider preference, visible to the next refresh."""
        self._preferred[namespace_id] = tuple(order)

    def set_permitted_operators(self, namespace_id: NamespaceId, operators: Iterable[OperatorId]) -> None:
        """Test helper: model a namespace config change to its operator gating, visible to the next grant."""
        self._permitted_operators[namespace_id] = frozenset(operators)

    def set_credentials(
        self, namespace_id: NamespaceId, capability_id: CapabilityId, credentials: Mapping[str, Any]
    ) -> None:
        self._credentials[(namespace_id, capability_id)] = dict(credentials)

File: typescript/src/orcastork/adapters/memory/catalog.ts
/**
 * In-memory `CapabilityCatalog` — explicit permitted-set + credentials maps.
 *
 * Enough to satisfy the port contract and the conformance suite; a deployment writes its own
 * adapter over its real configuration store. The mutators let tests model a namespace
 * enabling/revoking a capability between grants.
 *
 * @module
 */

import type { CapabilityId, NamespaceId, OperatorId } from '../../ids.js';
import type { CapabilityCatalog } from '../../ports/capability_catalog.js';

/** The secrets a capability is built from — whatever the deployment's catalog stores for it. */
export type Credentials = Readonly<Record<string, unknown>>;

/**
 * One namespace's credentials for one capability.
 *
 * An entry rather than a map key, because the Python catalog keys credentials by the
 * `(namespace, capability)` tuple and JavaScript has no value-equal composite key.
 */
export interface CredentialEntry {
  readonly namespaceId: NamespaceId;
  readonly capabilityId: CapabilityId;
  readonly credentials: Credentials;
}

/** What {@link InMemoryCapabilityCatalog} is seeded with; every part is optional and defaults to empty. */
export interface InMemoryCapabilityCatalogInit {
  readonly permitted?: Iterable<readonly [NamespaceId, Iterable<CapabilityId>]>;
  readonly credentials?: Iterable<CredentialEntry>;
  readonly preferred?: Iterable<readonly [NamespaceId, Iterable<CapabilityId>]>;
  readonly permittedOperators?: Iterable<readonly [NamespaceId, Iterable<OperatorId>]>;
}

/** Per-namespace configuration held in process memory. */
export class InMemoryCapabilityCatalog implements CapabilityCatalog {
  private readonly permitted = new Map<NamespaceId, ReadonlySet<CapabilityId>>();
  private readonly credentialsByNamespace = new Map<NamespaceId, Map<CapabilityId, Credentials>>();
  private readonly preferred = new Map<NamespaceId, readonly CapabilityId[]>();
  private readonly permittedOperatorsByNamespace = new Map<NamespaceId, ReadonlySet<OperatorId>>();

  public constructor(init: InMemoryCapabilityCatalogInit = {}) {
    for (const [namespaceId, capabilities] of init.permitted ?? []) {
      this.setPermitted(namespaceId, capabilities);
    }
    for (const entry of init.credentials ?? []) {
      this.setCredentials(entry.namespaceId, entry.capabilityId, entry.credentials);
    }
    for (const [namespaceId, order] of init.preferred ?? []) {
      this.setPreferredOrder(namespaceId, order);
    }
    for (const [namespaceId, operators] of init.permittedOperators ?? []) {
      this.setPermittedOperators(namespaceId, operators);
    }
  }

  public async permittedCapabilities(namespaceId: NamespaceId): Promise<ReadonlySet<CapabilityId>> {
    return this.permitted.get(namespaceId) ?? new Set<CapabilityId>();
  }

  public async permittedOperators(namespaceId: NamespaceId): Promise<ReadonlySet<OperatorId> | null> {
    // A namespace with no configured restriction is unrestricted (null), per the port contract — an
    // empty set is the explicit "run nothing" configuration, never the default.
    return this.permittedOperatorsByNamespace.get(namespaceId) ?? null;
  }

  public async credentials(namespaceId: NamespaceId, capabilityId: CapabilityId): Promise<Credentials | null> {
    // Hand back a copy so a caller mutating the returned mapping cannot alter catalog state.
    const stored = this.credentialsByNamespace.get(namespaceId)?.get(capabilityId);
    return stored === undefined ? null : { ...stored };
  }

  public async preferredOrder(namespaceId: NamespaceId): Promise<readonly CapabilityId[]> {
    return this.preferred.get(namespaceId) ?? [];
  }

  /** Test helper: model a namespace config change (enable/revoke) visible to the next grant. */
  public setPermitted(namespaceId: NamespaceId, capabilities: Iterable<CapabilityId>): void {
    this.permitted.set(namespaceId, new Set(capabilities));
  }

  /** Test helper: model a namespace changing its provider preference, visible to the next refresh. */
  public setPreferredOrder(namespaceId: NamespaceId, order: Iterable<CapabilityId>): void {
    this.preferred.set(namespaceId, [...order]);
  }

  /** Test helper: model a namespace config change to its operator gating, visible to the next grant. */
  public setPermittedOperators(namespaceId: NamespaceId, operators: Iterable<OperatorId>): void {
    this.permittedOperatorsByNamespace.set(namespaceId, new Set(operators));
  }

  /** Test helper: supply (or replace) the credentials one namespace holds for one capability. */
  public setCredentials(namespaceId: NamespaceId, capabilityId: CapabilityId, credentials: Credentials): void {
    const forNamespace = this.credentialsByNamespace.get(namespaceId) ?? new Map<CapabilityId, Credentials>();
    forNamespace.set(capabilityId, { ...credentials });
    this.credentialsByNamespace.set(namespaceId, forNamespace);
  }
}
